File: RotateArray_189.java
class Solution {
    public static void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        
        reverse(nums, 0, n - 1);
        reverse(nums, 0 , k - 1);
        reverse(nums, k, n - 1);
       
    }
    
    private static void reverse(int[] nums, int start, int end) {
        int temp;
        while(start < end) {
           temp = nums[start];
           nums[start] = nums[end];
           nums[end] = temp;
           start++;
           end--;
        }
    }
    public static void main(String args[]) {
        int[] nums = {1,2,3,4,5,6,7};
        rotate(nums, 3);
        for(int i = 0; i < nums.length; i++) {
            System.out.println(nums[i]);
        }

    }
}

File: RotateArray_189.py
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k = k % n

        self.reverse(nums, 0, n - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, n - 1)

        # k = k % len(nums)
        # nums[:] = nums[-k:] + nums[:-k]


    def reverse(self, nums, start, end):
        while start < end :
            temp = nums[start]
            nums[start] = nums[end]
            nums[end] = temp
            start += 1
            end -= 1

if __name__ == '__main__':
    slo = Solution()
    nums = [7,1,2,3,4,5,6]
    slo.rotate(nums,3)
    print(nums)